src/Algebra/Graph/AdjacencyMap/Algorithm.hs

```diff
@@ -212,19 +226,51 @@ dfs vs = concatMap flatten . dfsForestFrom vs
 reachable :: Ord a => a -> AdjacencyMap a -> [a]
 reachable x = concat . bfs [x]
 
+type ParentTable a = Map.Map a (Maybe a,Bool)
+data S a = S { table :: !(ParentTable a), order :: [a] }
+type TopOrder a = Either [a] [a]
+
+pattern TreeEdge  <- Nothing
+pattern BackEdge  <- Just (_,False)
+pattern OtherEdge <- Just (_,True)
+pattern Parent p  <- Just (Just p,_)
+
+topSort' :: (Ord a, MonadState (S a) m, MonadCont m)
+         => AdjacencyMap a -> m (TopOrder a)
+topSort' g = callCC $ \cyclic -> do
+  let unexplored u = gets (not . Map.member u . table)
+      parent u v = modify' (\(S p vs) -> S (Map.insert v (u,False) p) vs)
+      exit v = modify' (\(S p vs) -> S (Map.alter mark v p) (v:vs)) where
+        mark = fmap (fmap (const True)) -- mark tree as explored/done
+      edge_type v = gets (Map.lookup v . table)
+      retrace v vs@(u:_) table@(Map.lookup u -> ~(Parent w))
+        | v == u = vs
+        | v == w = v:vs
+        | otherwise = retrace v (w:vs) table
+      dfs u =
+        do forM_ (Set.toDescList $ postSet u g) $ \v ->
+             edge_type v >>= \case
+               TreeEdge  -> parent (Just u) v >> dfs v
+               BackEdge  -> cyclic . Left . retrace v [u] =<< gets table
+               OtherEdge -> pure ()
+           exit u
+  forM_ (map fst $ Map.toDescList $ adjacencyMap g) $
+    \v -> do new_tree <- unexplored v
+             when new_tree $ parent Nothing v >> dfs v
+  Right <$> gets order
+
 -- | Compute the /topological sort/ of a graph or return @Nothing@ if the graph
 -- is cyclic.
 --
 -- @
--- topSort (1 * 2 + 3 * 1)               == Just [3,1,2]
--- topSort (1 * 2 + 2 * 1)               == Nothing
+-- topSort (1 * 2 + 3 * 1)               == Right [3,1,2]
+-- topSort (1 * 2 + 2 * 1)               == Left [2,1]
```

Comment:
Is there a chance to return the lexicographically smallest cycle? In this case, it would be `[1,2]`.

Reply:
In the case that the `topSort` succeeds we get the smallest topological ordering. To get that, the biggest vertices are expanded first. As soon as a back edge is found, it is expanded and returned. It might be the lexicographically biggest cycle? 

I'm not sure if it would be possible to look for smaller cycles without losing the smallest topSort or earliest short circuit. I could potentially present the found cycle with the smallest element at the head of the list though?

Reply:
Not the lexicographically greatest:
```
λ> topSort (circuit [1..5] + circuit [5,4..1])
Left [5,4]
```

Reply:
Hmm, it looks more complicated than I thought. Perhaps, it's what happens in bipartite graphs:

https://github.com/snowleopard/alga/blob/00925932e060f005e2bcf463246762e34625f6f0/src/Algebra/Graph/Bipartite/AdjacencyMap.hs#L738-L742

That is, in this case, you found `[1, 2, 3, 4, 5]`, where the prefix is `[1, 2, 3]` and the cycle is `[4, 5]`. You just reported the cycle starting from the second vertex. If you report it from the first then I think we can use the same description as in the bipartite case.

Reply:
/cc @vasalf

Reply:
Hm! So I thought about it while running, and I believe it returns the lexicographically least cycle on `a` with opposite `Ord a` instance.

Reply:
Is it right that this `topSort` code goes through vertices in descending order and tries to find a cycle with a standard depth-first search? Then the returned cycle is not lexicographically largest.

For example, consider the following graph.

```Haskell
edges [(1,2),(2,1),(5,2),(3,4),(4,3)]
```

In this graph the algorithm which goes through vertices in descending order would first go from 5 and find the cycle `[2, 1]`, while there is a cycle `[4,3]`.

If I'm correct, `path ++ cycle` is lexicographically largest, just as in `detectParts` case. In my example, it is `[5] ++ [2,1]` versus `[] ++ [4,3]`.

Reply:
Ah yes, you're right. I think it is still the largest within in the connected component containing the largest vertex?

Reply:
You're characterization fails in the example `circuit [5,4..1] + circuit [1..5]`, where the lexicographically largest `path ++ cycle` would be `[5,4,3,2,1]`, but the cycle `[5,4]` is found by the algorithm. `[5,4]` being smallest with respect to `flip <`.

Reply:
Oh, there's a difference between «lexicographically largest» and «lexicographically smallest with respect to `flip <`» and now I see it.
Yes, I agree that `path ++ cycle` should be lexicographically smallest with respect to `flip <`.

Reply:
hah yes. I just noticed one more subtlety: taking your example, but with 5 -> 1 instead of 5 -> 2, we get
```
λ> topSort $ 1*2+2*1+3*4+4*3+5*1
Left (1 :| [2])
```
So it's lexicographically least with respect to flip <, up to rotation... I think